docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,343 @@
+kind: MachineDrainRule
+metadata:
+  name: portworx
+  namespace: default
+spec:
+  drain:
+    behavior: Drain
+    order: 100
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchLabels:
+        app: portworx
+```
+
+#### Drain labels
+
+We propose to introduce new labels to allow workload cluster admins/users to define drain behavior 
+for Pods. These labels would be either added directly to Pods or indirectly via Deployments, StatefulSets, etc.
+The labels will take precedence over `MachineDrainRules` specified in the management cluster.
+
+* `cluster.x-k8s.io/drain: skip`
+
+Initially we also considered adding a `cluster.x-k8s.io/drain-order` label. But we're not entirely sure about it 
+yet as someone who has access to the workload cluster (or maybe only specific Pods) would be able to influence the 
+drain order of the entire cluster, which might lead to problems. The skip label is safe in comparison because it 
+only influences the drain behavior of the Pod that has the label.
+
+#### Node drain behavior
+
+The following describes the new algorithm that decides which Pods should be drained and in which order.
+
+* List all Pods on the Node
+* For each Pod (first match applies):
+  * If the Pod is a DaemonSet Pod or a static Pod
+    * \=\> use `behavior: Skip`
+  * If the Pod has `cluster.x-k8s.io/drain: skip`
+    * \=\> use `behavior: Skip`
+  * If there is a matching `MachineDrainRule`
+    * \=\> use `behavior` and `order` from the first matching `MachineDrainRule` (based on alphabetical order)
+  * Otherwise:
+    * \=\> use `behavior: Drain` and `order: 0`
+* If there are no more Pods to be drained
+  * \=\> drain is completed
+* Else:
+  * \=\> evict all Pods with the lowest `order`, update condition and requeue
+
+Notes:
+
+* It does not make sense to evict DaemonSet Pods, because the DaemonSet controller always [adds tolerations for the 
+  unschedulable taint to all its Pods](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations)
+  so they would be just immediately rescheduled.
+* It does not make sense to evict static Pods, because the kubelet won’t actually terminate them and the Pod object 
+  will get recreated.
+* If multiple `MachineDrainRules` match the same Pod, the “first” rule (based on alphabetical order) will be applied
+* `Pods` with drain behavior `Drain` and given order will be evicted only after all other `Pods` with a lower order 
+  have been already deleted
+* When waiting for volume detach, volumes belonging to `Pods` with drain behavior `Skip` are going to be ignored.
+  Or in other words, Machine deletion can continue if only volumes belonging to `Pods` with drain behavior `Skip`
+  are still attached to the Node.
+
+#### Changes to wait for volume detach
+
+Today, after Node drain we are waiting for **all** volumes to be detached. We are going to change that behavior to ignore
+all attached volumes that belong to Pods for which we skipped the drain.
```

Comment:
What happens if the volume is not detached and the machine is shut down? I guess this varies per platform, but, are there ways that this could wedge CSI?